diff --git a/ioctlexact.c b/ioctlexact.c
index 09ee685..d222265 100644
--- a/ioctlexact.c
+++ b/ioctlexact.c
@@ -230,7 +230,14 @@ print_xen_domctl_enter(print_status *status, const struct xen_domctl *data)
 	XEN_DOMCTL(vm_event_op, xen_domctl_vm_event_op, vm_event_op)	// input
 #endif
 //	XEN_DOMCTL(mem_sharing_op, xen_domctl_mem_sharing_op, mem_sharing_op)	// depends
+#if defined(XEN_DOMCTL_disable_migrate)
 	XEN_DOMCTL(disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// input
+#elif defined(XEN_DOMCTL_set_disable_migrate)
+	XEN_DOMCTL(set_disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// input
+#endif
+#ifdef XEN_DOMCTL_query_disable_migrate
+	XEN_DOMCTL_OUT(query_disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// output
+#endif
 	XEN_DOMCTL_OUT(gettscinfo, xen_domctl_tsc_info, tsc_info)	// output
 	XEN_DOMCTL(settscinfo, xen_domctl_tsc_info, tsc_info)	// input
 	// TODO handle input and output of same field
@@ -340,7 +347,15 @@ print_xen_domctl_exit(print_status *status, const struct xen_domctl *data)
 	XEN_DOMCTL(vm_event_op, xen_domctl_vm_event_op, vm_event_op)	// input
 #endif
 //	XEN_DOMCTL(mem_sharing_op, xen_domctl_mem_sharing_op, mem_sharing_op)	// depends
+
+#if defined(XEN_DOMCTL_disable_migrate)
 	XEN_DOMCTL(disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// input
+#elif defined(XEN_DOMCTL_set_disable_migrate)
+	XEN_DOMCTL(set_disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// input
+#endif
+#ifdef XEN_DOMCTL_query_disable_migrate
+	XEN_DOMCTL_OUT(query_disable_migrate, xen_domctl_disable_migrate, disable_migrate)	// output
+#endif
 	XEN_DOMCTL_OUT(gettscinfo, xen_domctl_tsc_info, tsc_info)	// output
 	XEN_DOMCTL(settscinfo, xen_domctl_tsc_info, tsc_info)	// input
 	// TODO handle input and output of same field
