diff --git a/ioctlexact.c b/ioctlexact.c
index d222265..a8bc0cd 100644
--- a/ioctlexact.c
+++ b/ioctlexact.c
@@ -54,13 +54,13 @@ static const switch_item privcmd_hvmop_cmd_items[] = {
 	HVMOP(set_param, xen_hvm_param, 0)
 // TODO output !!!
 	HVMOP(get_param, xen_hvm_param, 1)
-	HVMOP(set_pci_intx_level, xen_hvm_set_pci_intx_level, 0)
-	HVMOP(set_isa_irq_level, xen_hvm_set_isa_irq_level, 0)
-	HVMOP(set_pci_link_route, xen_hvm_set_pci_link_route, 0)
+	/* HVMOP(set_pci_intx_level, xen_hvm_set_pci_intx_level, 0) */
+	/* HVMOP(set_isa_irq_level, xen_hvm_set_isa_irq_level, 0) */
+	/* HVMOP(set_pci_link_route, xen_hvm_set_pci_link_route, 0) */
 	HVMOP(flush_tlbs, void, 0)
 // HVMOP_track_dirty_vram
-	HVMOP(modified_memory, xen_hvm_modified_memory, 0)
-	HVMOP(set_mem_type, xen_hvm_set_mem_type, 0)
+	/* HVMOP(modified_memory, xen_hvm_modified_memory, 0) */
+	/* HVMOP(set_mem_type, xen_hvm_set_mem_type, 0) */
 	HVMOP(process_dying, void, 0)
 	HVMOP(pagetable_dying, xen_hvm_pagetable_dying, 0)
 // TODO output !!!
@@ -69,7 +69,7 @@ static const switch_item privcmd_hvmop_cmd_items[] = {
 	HVMOP(set_mem_access, xen_hvm_set_mem_access, 0)
 // TODO output !!!
 	HVMOP(get_mem_access, xen_hvm_get_mem_access, 1)
-	HVMOP(inject_trap, xen_hvm_inject_trap, 0)
+	/* HVMOP(inject_trap, xen_hvm_inject_trap, 0) */
 // TODO output !!!
 	HVMOP(get_mem_type, xen_hvm_get_mem_type, 1)
 	HVMOP(set_driver_version, xen_hvm_set_driver_version, 0)
diff --git a/structures.h b/structures.h
index a9d8dab..13994c6 100644
--- a/structures.h
+++ b/structures.h
@@ -69,6 +69,7 @@ START_STRUCT(privcmd_hypercall)
 END_STRUCT(privcmd_hypercall)
 #undef STRUCT
 
+#if 0
 #define STRUCT struct xen_hvm_set_pci_intx_level
 START_STRUCT(xen_hvm_set_pci_intx_level)
 	BASE_ITEM(domid, domid)
@@ -95,6 +96,7 @@ START_STRUCT(xen_hvm_set_pci_link_route)
 	BASE_ITEM(u8, isa_irq)
 END_STRUCT(xen_hvm_set_pci_link_route)
 #undef STRUCT
+#endif
 
 #define STRUCT struct xen_hvm_param
 START_STRUCT(xen_hvm_param)
@@ -104,6 +106,7 @@ START_STRUCT(xen_hvm_param)
 END_STRUCT(xen_hvm_param)
 #undef STRUCT
 
+#if 0
 #define STRUCT struct xen_hvm_modified_memory
 START_STRUCT(xen_hvm_modified_memory)
 	BASE_ITEM(domid, domid)
@@ -120,6 +123,7 @@ START_STRUCT(xen_hvm_set_mem_type)
 	BASE_ITEM(u64, first_pfn)
 END_STRUCT(xen_hvm_set_mem_type)
 #undef STRUCT
+#endif
 
 #define STRUCT struct xen_hvm_pagetable_dying
 FIXED_ARRAY(xen_hvm_pagetable_dying_pads, pad, u16)
@@ -164,6 +168,7 @@ END_STRUCT(xen_hvm_get_mem_access)
 #undef STRUCT
 #endif /* XEN_SYSCTL_INTERFACE_VERSION < 0x0000000B */
 
+#if 0
 #define STRUCT struct xen_hvm_inject_trap
 START_STRUCT(xen_hvm_inject_trap)
 	BASE_ITEM(domid, domid)
@@ -172,6 +177,7 @@ START_STRUCT(xen_hvm_inject_trap)
 	BASE_ITEM(u64, cr2)
 END_STRUCT(xen_hvm_inject_trap)
 #undef STRUCT
+#endif
 
 #define STRUCT struct xen_hvm_get_mem_type
 FIXED_ARRAY(xen_hvm_get_mem_type_pads, pad, u16)
