diff --git a/structures.h b/structures.h
index b22da3b..473c6e4 100644
--- a/structures.h
+++ b/structures.h
@@ -137,6 +137,15 @@ START_STRUCT(xen_hvm_get_time)
 END_STRUCT(xen_hvm_get_time)
 #undef STRUCT
 
+#if XEN_SYSCTL_INTERFACE_VERSION < 0x0000000B
+/*
+ * Gross hack, to counter a gross API breakage.  Xen 4.5 deprecated the
+ * HVMOP_{get,set}_mem_access hypercalls, removing their structure definitions
+ * but not hypercall numbers.
+ *
+ * While the sysctl interface version is unrelated to hvmops, it serves as a
+ * Xen-4.4 vs Xen-4.5 gate.
+ */
 #define STRUCT struct xen_hvm_set_mem_access
 START_STRUCT(xen_hvm_set_mem_access)
 	BASE_ITEM(domid, domid)
@@ -153,6 +162,7 @@ START_STRUCT(xen_hvm_get_mem_access)
 	BASE_ITEM(u64, pfn)
 END_STRUCT(xen_hvm_get_mem_access)
 #undef STRUCT
+#endif /* XEN_SYSCTL_INTERFACE_VERSION < 0x0000000B */
 
 #define STRUCT struct xen_hvm_inject_trap
 START_STRUCT(xen_hvm_inject_trap)
@@ -293,7 +303,15 @@ END_STRUCT(xenctl_bitmap)
 #define STRUCT struct xen_domctl_vcpuaffinity
 START_STRUCT(xen_domctl_vcpuaffinity)
 	BASE_ITEM(u32, vcpu)
+#ifndef XEN_VCPUAFFINITY_HARD
+/* Xen 4.4 and older.  Just a single bitmap */
 	BASE_ITEM(xenctl_bitmap, cpumap)
+#else
+/* Xen 4.5 - Added a soft vs hard split */
+	BASE_ITEM(u32, flags)
+	BASE_ITEM(xenctl_bitmap, cpumap_hard)
+	BASE_ITEM(xenctl_bitmap, cpumap_soft)
+#endif
 END_STRUCT(xen_domctl_vcpuaffinity)
 #undef STRUCT
 
